pkg/logutil/zap_text_encoder.go

```diff
@@ -0,0 +1,572 @@
+		final.endQuoteFiled()
+	}
+
+	if final.LineEnding != "" {
+		final.buf.AppendString(final.LineEnding)
+	} else {
+		final.buf.AppendString(zapcore.DefaultLineEnding)
+	}
+
+	ret := final.buf
+	putTextEncoder(final)
+	return ret, nil
+}
+
+func (enc *textEncoder) truncate() {
+	enc.buf.Reset()
+}
+
+func (enc *textEncoder) closeOpenNamespaces() {
+	for i := 0; i < enc.openNamespaces; i++ {
+		enc.buf.AppendByte('}')
+	}
+}
+
+func (enc *textEncoder) addKey(key string) {
+	enc.addElementSeparator()
+	enc.safeAddStringWithQuote(key)
+	enc.buf.AppendByte('=')
+}
+
+func (enc *textEncoder) addElementSeparator() {
+	last := enc.buf.Len() - 1
+	if last < 0 {
+		return
+	}
+	switch enc.buf.Bytes()[last] {
+	case '{', '[', ':', ',', ' ', '=':
+		return
+	default:
+		enc.buf.AppendByte(',')
+	}
+}
+
+func (enc *textEncoder) appendFloat(val float64, bitSize int) {
+	enc.addElementSeparator()
+	switch {
+	case math.IsNaN(val):
+		enc.buf.AppendString(`"NaN"`)
+	case math.IsInf(val, 1):
+		enc.buf.AppendString(`"+Inf"`)
+	case math.IsInf(val, -1):
+		enc.buf.AppendString(`"-Inf"`)
+	default:
+		enc.buf.AppendFloat(val, bitSize)
+	}
+}
+
+// safeAddString JSON-escapes a string and appends it to the internal buffer.
+// Unlike the standard library's encoder, it doesn't attempt to protect the
+// user from browser vulnerabilities or JSONP-related problems.
+func (enc *textEncoder) safeAddString(s string) {
+	for i := 0; i < len(s); {
+		if enc.tryAddRuneSelf(s[i]) {
+			i++
+			continue
+		}
+		r, size := utf8.DecodeRuneInString(s[i:])
+		if enc.tryAddRuneError(r, size) {
+			i++
+			continue
+		}
+		enc.buf.AppendString(s[i : i+size])
+		i += size
+	}
+}
+
+// safeAddStringWithQuote will automatically add quotoes.
+func (enc *textEncoder) safeAddStringWithQuote(s string) {
+	if !enc.needDoubleQuotes(s) {
+		enc.buf.AppendString(s)
+		return
+	}
+	enc.buf.AppendByte('"')
+	enc.safeAddString(s)
+	enc.buf.AppendByte('"')
+}
+
+// safeAddByteString is no-alloc equivalent of safeAddString(string(s)) for s []byte.
+func (enc *textEncoder) safeAddByteString(s []byte) {
+	for i := 0; i < len(s); {
+		if enc.tryAddRuneSelf(s[i]) {
+			i++
+			continue
+		}
+		r, size := utf8.DecodeRune(s[i:])
+		if enc.tryAddRuneError(r, size) {
+			i++
+			continue
+		}
+		enc.buf.Write(s[i : i+size])
+		i += size
+	}
+}
+
+// For https://github.com/tikv/rfcs/blob/master/text/2018-12-19-unified-log-format.md#log-fields-section.
+func (enc *textEncoder) needDoubleQuotes(s string) bool {
+	for i := 0; i < len(s); {
+		b := s[i]
+		if b <= 0x20 {
+			return true
+		}
+		switch b {
+		case '\\', '"', '[', ']', '=':
+			return true
+		case '\n', '\t', '\r':
+			return true
+		}
+		if b >= utf8.RuneSelf {
+			return true
+		}
+		i++
+	}
+	return false
+}
+
+// tryAddRuneSelf appends b if it is valid UTF-8 character represented in a single byte.
+func (enc *textEncoder) tryAddRuneSelf(b byte) bool {
+	if b >= utf8.RuneSelf {
+		return false
+	}
+	if 0x20 <= b && b != '\\' && b != '"' {
```

Comment:
I guess you missed to escape `[` and  `]` here.

Reply:
`[` and `]` do not need to escape.

Reply:
According to [RFC](https://github.com/tikv/rfcs/blob/master/text/2018-12-19-unified-log-format.md#log-fields-section), them should be escaped. 
> U+005B (LEFT SQUARE BRACKET)
> U+005D (RIGHT SQUARE BRACKET)

Reply:
@nolouch is correct. JSON won't escape these characters.